crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2574,10 +2582,7 @@ impl<'db> TypeInferenceBuilder<'db> {
                 self.infer_call_expression(call_expr);
                 Type::Unknown
             }
-            ast::Expr::FString(fstring) => {
-                self.infer_fstring_expression(fstring);
-                Type::Unknown
-            }
+            ast::Expr::FString(_) => Type::Unknown,
```

Comment:
Is this change intentional?

Reply:
I think - as far as I understand, `infer_fstring_expression` does not mutate anything in self, so calling it without using the returned value does nothing, doesn't it?
As far as the return we have (`Type::Unknown`), it should still be this as ` f-strings` are invalid in annotation context (to my knowledge)

Reply:
We maintain an invariant that all expressions in a scope are visited when inferring types in that scope. Since f-strings can have arbitrary sub-expressions, failing to call `infer_fstring_expression` here can mean that we fail to visit those sub-expressions or assign them a type (and potentially miss diagnostics in them.)

There will be some decisions to make at some point around how we want to handle invalid expression kinds in annotations -- should we still emit internal diagnostics for them, or just emit the "not valid in annotations" diagnostic, assign type unknown, and move on? I don't have a clear answer on that right now, but for this PR, for now, I think this change should be reverted so that we maintain our "visit all expressions" invariant.

Reply:
> We maintain an invariant that all expressions in a scope are visited when inferring types in that scope. 

That's very interesting to know. I think this applies to my PR as there is (in the latest) version a short-circuit logic to stop iterating over `f-string` parts (that could be arbitrarily long) once we know for sure the type to infer (either `str` instance, or `StringLiteral`).

Should I go back and ensure that we do call `infer_expression` even when we already figured out the return type?